beginner_source/nestedtensor_tutorial.py

```diff
@@ -0,0 +1,449 @@
+"""
+Nested Tensors
+===============================================================
+
+Nested tensor is very similar to regular tensor, except for the shape:
+
+* for a regular tensor, each dimension has a size
+
+* for a nested tensor, not all dimensions have regular sizes: some of them are jagged
+
+Nested tensor arises from many application scenarios:
+
+* in NLP, sentenses have variable lengthes, so a batch of sentenses forms a nested tensor
+
+* in CV, images have variable shapes, so a batch of images forms a nested tensor
+
+For practical reason, conceptually we implement nested tensor
+as a batch of tensors with different shapes,
+i.e. dimension 0 is assumed to be the batch dimension.
+"""
+
+import torch
+import torch.nn.functional as F
+
+device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
+
+######################################################################
+# Nested Tensor Initialization
+# ----------------
+#
+
+######################################################################
+# From python frontend, for now nested tensor can only be created from a list of tensors.
+nt = torch.nested_tensor([torch.randn((2, 6)), torch.randn((3, 6))], device=device)
+print(nt)
+
+######################################################################
+# Nested Tensor Operations
+# ----------------
+#
+
+######################################################################
+# Operation coverage for nested tensor is narrower than regular tensors.
+# For now only basic operations such as index, dropout, softmax, transpose, reshape, linear, bmm are covered.
+# If you need certain operations, please file an `issue <https://github.com/pytorch/nestedtensor>`__.
+#
+# **reshape**
+#
+# Note that -1 no longer infers as the case of regular tensor.
+# Instead, it inherits the old size (here 2 for ``nt[0]`` and 3 for ``nt[1]``).
+# -1 is the only legal size to specify for a jagged dimension.
+nt1 = nt.reshape(2, -1, 2, 3)
+print(nt1)
+
+######################################################################
+# **transpose**
+#
+# Note that transposing dimension 0 is forbidden,
+# i.e. the batch dimension is special and should always be dimension 0.
+nt2 = nt1.transpose(1, 2)
+print(nt2)
+
+######################################################################
+# **others**
+#
+# Other operations have same semantics as regular tensors.
+nt_mm = torch.nested_tensor([torch.randn((2, 3, 4)), torch.randn((2, 3, 5))], device=device)
+nt3 = torch.matmul(nt2, nt_mm)
+print(nt3)
+
+######################################################################
+# Why Nested Tensor
```

Comment:
nit: I've usually used `NestedTensor` for the name of the data structure, `nested_tensor` for the factory function and `nestedtensor` for the name of the project. See

Reply:
Is there a link to some reference?

Reply:
I was going to add something, but now I can't find it anymore.